diff -ruNa labplot-2.11.1.orig/src/backend/datasources/filters/AsciiFilter.cpp labplot-2.11.1/src/backend/datasources/filters/AsciiFilter.cpp
--- labplot-2.11.1.orig/src/backend/datasources/filters/AsciiFilter.cpp	2024-07-15 22:50:37.000000000 +0200
+++ labplot-2.11.1/src/backend/datasources/filters/AsciiFilter.cpp	2024-11-13 22:19:02.989697570 +0100
@@ -665,8 +665,13 @@
 		DEBUG(Q_FUNC_INFO << ", column mode (after checking more lines) = " << ENUM_TO_STRING(AbstractColumn, ColumnMode, mode));
 #endif
 
+	// if header enabled: increase max lines (not if size_t::max())
+	size_t maxLinesToRead = maxLines;
+	if (headerEnabled && maxLines < std::numeric_limits<std::size_t>::max())
+		maxLinesToRead++;	
+
 	// ATTENTION: This resets the position in the device to 0
-	m_actualRows = (int)q->lineNumber(device, maxLines);
+	m_actualRows = (int)q->lineNumber(device, maxLinesToRead);
 	DEBUG(Q_FUNC_INFO << ", m_actualRows: " << m_actualRows << ", startRow (after header): " << startRow << ", endRow: " << endRow)
 
 	DEBUG(Q_FUNC_INFO << ", headerEnabled = " << headerEnabled << ", headerLine = " << headerLine << ", m_actualStartRow = " << m_actualStartRow)
